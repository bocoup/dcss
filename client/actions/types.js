export const LOG_IN = 'LOG_IN';
export const LOG_OUT = 'LOG_OUT';

export const GET_CATEGORIES_SUCCESS = 'GET_CATEGORIES_SUCCESS';
export const GET_CATEGORIES_ERROR = 'GET_CATEGORIES_ERROR';

export const SET_USERS = 'SET_USERS';

export const GET_USER = 'GET_USER';
export const GET_USER_SUCCESS = 'GET_USER_SUCCESS';
export const GET_USER_ERROR = 'GET_USER_ERROR';

export const GET_SESSION_SUCCESS = 'GET_SESSION_SUCCESS';
export const GET_SESSION_ERROR = 'GET_SESSION_ERROR';

export const SET_USER = 'SET_USER';
export const SET_USER_SUCCESS = 'SET_USER_SUCCESS';
export const SET_USER_ERROR = 'SET_USER_ERROR';

export const GET_USERS = 'GET_USERS';
export const GET_USERS_SUCCESS = 'GET_USERS_SUCCESS';
export const GET_USERS_ERROR = 'GET_USERS_ERROR';

export const GET_RESPONSE = 'GET_RESPONSE';
export const GET_RESPONSE_SUCCESS = 'GET_RESPONSE_SUCCESS';
export const GET_RESPONSE_ERROR = 'GET_RESPONSE_ERROR';

export const GET_TRANSCRIPT = 'GET_TRANSCRIPT';
export const GET_TRANSCRIPT_SUCCESS = 'GET_TRANSCRIPT_SUCCESS';
export const GET_TRANSCRIPT_ERROR = 'GET_TRANSCRIPT_ERROR';

export const GET_RESPONSES = 'GET_RESPONSES';
export const GET_RESPONSES_SUCCESS = 'GET_RESPONSES_SUCCESS';
export const GET_RESPONSES_ERROR = 'GET_RESPONSE_ERROR';

export const SET_RESPONSE = 'SET_RESPONSE';
export const SET_RESPONSE_SUCCESS = 'SET_RESPONSE_SUCCESS';
export const SET_RESPONSE_ERROR = 'SET_RESPONSE_ERROR';

export const SET_RESPONSES = 'SET_RESPONSES';
export const SET_RESPONSES_SUCCESS = 'SET_RESPONSES_SUCCESS';
export const SET_RESPONSES_ERROR = 'SET_RESPONSES_ERROR';

export const GET_RUN = 'GET_RUN';
export const GET_RUN_SUCCESS = 'GET_RUN_SUCCESS';
export const GET_RUN_ERROR = 'GET_RUN_ERROR';

export const GET_RUNS = 'GET_RUNS';
export const GET_RUNS_SUCCESS = 'GET_RUNS_SUCCESS';
export const GET_RUNS_ERROR = 'GET_RUNS_ERROR';

export const GET_RUN_DATA = 'GET_RUN_DATA';
export const GET_RUN_DATA_SUCCESS = 'GET_RUN_DATA_SUCCESS';
export const GET_RUN_DATA_ERROR = 'GET_RUN_DATA_ERROR';

export const SET_USER_ROLE_SUCCESS = 'SET_USER_ROLE_SUCCESS';
export const SET_USER_ROLE_ERROR = 'SET_USER_ROLE_ERROR';

export const SET_RUN = 'SET_RUN';
export const SET_RUN_SUCCESS = 'SET_RUN_SUCCESS';
export const SET_RUN_ERROR = 'SET_RUN_ERROR';

export const DELETE_SCENARIO_SUCCESS = 'DELETE_SCENARIO_SUCCESS';
export const DELETE_SCENARIO_ERROR = 'DELETE_SCENARIO_ERROR';

export const UNLOCK_SCENARIO_SUCCESS = 'UNLOCK_SCENARIO_SUCCESS';
export const UNLOCK_SCENARIO_ERROR = 'UNLOCK_SCENARIO_ERROR';

export const GET_SCENARIO = 'GET_SCENARIO';
export const GET_SCENARIO_SUCCESS = 'GET_SCENARIO_SUCCESS';
export const GET_SCENARIO_ERROR = 'GET_SCENARIO_ERROR';

export const GET_SCENARIOS = 'GET_SCENARIOS';
export const GET_SCENARIOS_SUCCESS = 'GET_SCENARIOS_SUCCESS';
export const GET_SCENARIOS_ERROR = 'GET_SCENARIOS_ERROR';

export const SET_SCENARIO = 'SET_SCENARIO';
export const SET_SCENARIO_SUCCESS = 'SET_SCENARIO_SUCCESS';
export const SET_SCENARIO_ERROR = 'SET_SCENARIO_ERROR';

export const GET_RUN_HISTORY = 'GET_RUN_HISTORY';
export const GET_RUN_HISTORY_SUCCESS = 'GET_RUN_HISTORY_SUCCESS';
export const GET_RUN_HISTORY_ERROR = 'GET_RUN_HISTORY_ERROR';

export const SET_SCENARIOS = 'SET_SCENARIOS';
export const SET_SCENARIOS_SUCCESS = 'SET_SCENARIOS_SUCCESS';
export const SET_SCENARIOS_ERROR = 'SET_SCENARIOS_ERROR';

export const DELETE_SLIDE_SUCCESS = 'DELETE_SLIDE_SUCCESS';
export const DELETE_SLIDE_ERROR = 'DELETE_SLIDE_ERROR';

export const SET_SLIDES = 'SET_SLIDES';

export const GET_SLIDES = 'GET_SLIDES';
export const GET_SLIDES_SUCCESS = 'GET_SLIDES_SUCCESS';
export const GET_SLIDES_ERROR = 'GET_SLIDES_ERROR';

export const CREATE_COHORT = 'CREATE_COHORT';
export const CREATE_COHORT_SUCCESS = 'CREATE_COHORT_SUCCESS';
export const CREATE_COHORT_ERROR = 'CREATE_COHORT_ERROR';

export const SET_COHORT = 'SET_COHORT';
export const SET_COHORT_SUCCESS = 'SET_COHORT_SUCCESS';
export const SET_COHORT_ERROR = 'SET_COHORT_ERROR';

export const SET_COHORT_USER_ROLE = 'SET_COHORT_USER_ROLE';
export const SET_COHORT_USER_ROLE_SUCCESS = 'SET_COHORT_USER_ROLE_SUCCESS';
export const SET_COHORT_USER_ROLE_ERROR = 'SET_COHORT_USER_ROLE_ERROR';

export const GET_COHORT = 'GET_COHORT';
export const GET_COHORT_SUCCESS = 'GET_COHORT_SUCCESS';
export const GET_COHORT_ERROR = 'GET_COHORT_ERROR';

export const GET_COHORT_DATA = 'GET_COHORT_DATA';
export const GET_COHORT_RUN_DATA_SUCCESS = 'GET_COHORT_RUN_DATA_SUCCESS';
export const GET_COHORT_RUN_DATA_ERROR = 'GET_COHORT_RUN_DATA_ERROR';

export const GET_COHORT_PARTICIPANTS = 'GET_COHORT_PARTICIPANTS';
export const GET_COHORT_PARTICIPANTS_SUCCESS =
  'GET_COHORT_PARTICIPANTS_SUCCESS';
export const GET_COHORT_PARTICIPANTS_ERROR = 'GET_COHORT_PARTICIPANTS_ERROR';

export const LINK_RUN_TO_COHORT = 'LINK_RUN_TO_COHORT';
export const LINK_RUN_TO_COHORT_SUCCESS = 'LINK_RUN_TO_COHORT_SUCCESS';
export const LINK_RUN_TO_COHORT_ERROR = 'LINK_RUN_TO_COHORT_ERROR';

export const GET_ALL_COHORTS = 'GET_ALL_COHORTS';
export const GET_ALL_COHORTS_SUCCESS = 'GET_ALL_COHORTS_SUCCESS';
export const GET_ALL_COHORTS_ERROR = 'GET_ALL_COHORTS_ERROR';

export const GET_USER_COHORTS = 'GET_USER_COHORTS';
export const GET_USER_COHORTS_SUCCESS = 'GET_USER_COHORTS_SUCCESS';
export const GET_USER_COHORTS_ERROR = 'GET_USER_COHORTS_ERROR';
