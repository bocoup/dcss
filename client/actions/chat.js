import {
  GET_CHAT_ERROR,
  GET_CHAT_SUCCESS,
  GET_CHATS_ERROR,
  GET_CHATS_SUCCESS,
  SET_CHAT_ERROR,
  SET_CHAT_SUCCESS,
  LINK_CHAT_TO_RUN_ERROR,
  LINK_CHAT_TO_RUN_SUCCESS
} from './types';

export let getChat = id => async dispatch => {
  try {
    const res = await (await fetch(`/api/chats/${id}`)).json();

    if (res.error) {
      throw res;
    }
    const { chat } = res;
    dispatch({ type: GET_CHAT_SUCCESS, chat });
    return chat;
  } catch (error) {
    dispatch({ type: GET_CHAT_ERROR, error });
    return null;
  }
};

export let setChat = (id, params) => async dispatch => {
  try {
    if (Object.values(params).length) {
      const body = JSON.stringify(params);
      const res = await (await fetch(`/api/chats/${id}`, {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json'
        },
        body
      })).json();

      if (res.error) {
        throw res;
      }

      const { chat } = res;
      dispatch({ type: SET_CHAT_SUCCESS, chat });
      return chat;
    }
    return null;
  } catch (error) {
    dispatch({ type: SET_CHAT_ERROR, error });
    return null;
  }
};

export let createChat = params => async dispatch => {
  try {
    if (Object.values(params).length) {
      const { lobby_id } = params;
      const body = JSON.stringify({
        lobby_id
      });
      const res = await (await fetch(`/api/chats`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body
      })).json();

      if (res.error) {
        throw res;
      }
      const { chat } = res;
      dispatch({ type: GET_CHAT_SUCCESS, chat });
      return chat;
    }
    return null;
  } catch (error) {
    dispatch({ type: GET_CHAT_ERROR, error });
    return null;
  }
};

export let getChats = () => async dispatch => {
  try {
    const res = await (await fetch('/api/chats')).json();

    if (res.error) {
      throw res;
    }

    const { chats = [] } = res;
    dispatch({ type: GET_CHATS_SUCCESS, chats });
    return chats;
  } catch (error) {
    dispatch({ type: GET_CHATS_ERROR, error });
    return null;
  }
};

export let linkChatToRun = (chat_id, run_id) => async dispatch => {
  try {
    const res = await (await fetch(
      `/api/chats/link/${chat_id}/run/${run_id}`
    )).json();

    if (res.error) {
      throw res;
    }
    const { chat } = res;
    dispatch({ type: LINK_CHAT_TO_RUN_SUCCESS, chat });
    return chat;
  } catch (error) {
    dispatch({ type: LINK_CHAT_TO_RUN_ERROR, error });
    return null;
  }
};
