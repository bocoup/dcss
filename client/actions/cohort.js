import {
  // CREATE_COHORT,
  CREATE_COHORT_SUCCESS,
  CREATE_COHORT_ERROR,
  SET_COHORT_SUCCESS,
  SET_COHORT_ERROR,
  // GET_COHORT,
  GET_COHORT_SUCCESS,
  GET_COHORT_ERROR,
  // GET_COHORT_PARTICIPANTS,
  GET_COHORT_PARTICIPANTS_SUCCESS,
  GET_COHORT_PARTICIPANTS_ERROR,
  // GET_COHORT_RUN_DATA_SUCCESS,
  GET_COHORT_RUN_DATA_ERROR,
  // GET_ALL_COHORTS,
  GET_ALL_COHORTS_SUCCESS,
  GET_ALL_COHORTS_ERROR,
  // GET_USER_COHORTS,
  GET_USER_COHORTS_SUCCESS,
  GET_USER_COHORTS_ERROR,
  // LINK_COHORT_TO_RUN,
  LINK_RUN_TO_COHORT_SUCCESS,
  LINK_RUN_TO_COHORT_ERROR,
  // SET_COHORT_USER_ROLE,
  SET_COHORT_USER_ROLE_SUCCESS,
  SET_COHORT_USER_ROLE_ERROR
} from './types';

export const createCohort = ({ name }) => async dispatch => {
  try {
    const res = await (await fetch('/api/cohort', {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ name })
    })).json();

    if (res.error) {
      throw res;
    }
    const { cohort } = res;

    dispatch({ type: CREATE_COHORT_SUCCESS, cohort });
    // return the cohort to the promise action for redirection purposes
    return cohort;
  } catch (error) {
    dispatch({ type: CREATE_COHORT_ERROR, error });
    return null;
  }
};

export const setCohort = cohort => async dispatch => {
  // dispatch({ type: SET_COHORT, cohort });
  try {
    const { scenarios } = cohort;
    const body = JSON.stringify({
      scenarios
    });
    await (await fetch(`/api/cohort/${cohort.id}/scenarios`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body
    })).json();
    dispatch({ type: SET_COHORT_SUCCESS, cohort });
  } catch (error) {
    dispatch({ type: SET_COHORT_ERROR, error });
    return null;
  }
};

export const getCohort = id => async dispatch => {
  if (Number.isNaN(id)) {
    return;
  }
  try {
    const res = await (await fetch(`/api/cohort/${id}`)).json();
    if (res.error) {
      throw res;
    }
    const { cohort } = res;
    dispatch({ type: GET_COHORT_SUCCESS, cohort });
    // return the cohort to the promise action for redirection purposes
    return cohort;
  } catch (error) {
    dispatch({ type: GET_COHORT_ERROR, error });
    return null;
  }
};

export const getCohorts = () => async dispatch => {
  try {
    const response = await (await fetch('/api/cohort/my')).json();

    if (response.error) {
      throw response;
    }
    const { cohorts } = response;
    dispatch({ type: GET_USER_COHORTS_SUCCESS, cohorts });
  } catch (error) {
    dispatch({ type: GET_USER_COHORTS_ERROR, error });
    return null;
  }
};

export const getAllCohorts = () => async dispatch => {
  try {
    const res = await (await fetch('/api/cohort/all')).json();
    if (res.error) {
      throw res;
    }
    const { cohorts } = res;
    dispatch({ type: GET_ALL_COHORTS_SUCCESS, cohorts });
    return cohorts;
  } catch (error) {
    dispatch({ type: GET_ALL_COHORTS_ERROR, error });
    return null;
  }
};

export const getCohortParticipants = id => async dispatch => {
  try {
    const res = await (await fetch(`/api/cohort/${id}`)).json();

    if (res.error) {
      throw res;
    }
    const {
      cohort,
      cohort: { users }
    } = res;
    // Dispatch the entire "cohort", but only the "users" property
    // will be used in the reducer.
    dispatch({ type: GET_COHORT_PARTICIPANTS_SUCCESS, cohort });
    // return the cohort to the promise action for redirection purposes
    return users;
  } catch (error) {
    dispatch({ type: GET_COHORT_PARTICIPANTS_ERROR, error });
    return null;
  }
};

export const getCohortData = (
  cohort_id,
  participant_id,
  scenario_id
) => async dispatch => {
  try {
    const endpoint = scenario_id
      ? `/api/cohort/${cohort_id}/scenario/${scenario_id}`
      : `/api/cohort/${cohort_id}/participant/${participant_id}`;

    const res = await (await fetch(endpoint)).json();

    if (res.error) {
      throw res;
    }

    const { prompts, responses } = res;

    // This dispatch does not update internal state, since
    // such state would be thrown away:
    // dispatch({ type: GET_COHORT_RUN_DATA_SUCCESS, prompts, responses });
    //
    return { prompts, responses };
  } catch (error) {
    dispatch({ type: GET_COHORT_RUN_DATA_ERROR, error });
    return null;
  }
};

export const linkRunToCohort = (cohort_id, run_id) => async dispatch => {
  try {
    const res = await (await fetch(
      `/api/cohort/${cohort_id}/run/${run_id}`
    )).json();
    if (res.error) {
      throw res;
    }
    const { cohort } = res;
    dispatch({ type: LINK_RUN_TO_COHORT_SUCCESS });
    dispatch({ type: GET_COHORT_SUCCESS, cohort });
    return cohort;
  } catch (error) {
    dispatch({ type: LINK_RUN_TO_COHORT_ERROR, error });
    return null;
  }
};
// This is used to
export const linkUserToCohort = (cohort_id, role) => async dispatch => {
  try {
    const users = await (await fetch(
      `/api/cohort/${cohort_id}/join/${role}`
    )).json();
    if (users.error) {
      throw users;
    }
    dispatch({ type: SET_COHORT_USER_ROLE_SUCCESS, users });
  } catch (error) {
    dispatch({ type: SET_COHORT_USER_ROLE_ERROR, error });
    return null;
  }
};

export const addCohortUserRole = (
  cohort_id,
  user_id,
  role
) => async dispatch => {
  try {
    const body = JSON.stringify({
      cohort_id,
      user_id,
      roles: [role]
    });
    const result = await (await fetch('/api/cohort/${cohort_id}/roles/add', {
      headers: {
        'Content-Type': 'application/json'
      },
      method: 'POST',
      body
    })).json();

    const { cohort } = result;
    dispatch({ type: SET_COHORT_USER_ROLE_SUCCESS });
    dispatch({ type: GET_COHORT_SUCCESS, cohort });
    return result;
  } catch (error) {
    dispatch({ type: SET_COHORT_USER_ROLE_ERROR, error });
    return null;
  }
};

export const deleteCohortUserRole = (
  cohort_id,
  user_id,
  role
) => async dispatch => {
  try {
    const body = JSON.stringify({
      cohort_id,
      user_id,
      roles: [role]
    });
    const result = await (await fetch('/api/cohort/${cohort_id}/roles/delete', {
      headers: {
        'Content-Type': 'application/json'
      },
      method: 'POST',
      body
    })).json();
    const { cohort } = result;
    dispatch({ type: SET_COHORT_USER_ROLE_SUCCESS });
    dispatch({ type: GET_COHORT_SUCCESS, cohort });
    return result;
  } catch (error) {
    dispatch({ type: SET_COHORT_USER_ROLE_ERROR, error });
    return null;
  }
};
