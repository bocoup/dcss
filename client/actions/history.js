import { GET_RUN_HISTORY_SUCCESS, GET_RUN_HISTORY_ERROR } from './types';

export const getHistoryForScenario = (
  scenario_id,
  cohort_id
) => async dispatch => {
  try {
    const endpoint = cohort_id
      ? `/api/history/${scenario_id}/cohort/${cohort_id}`
      : `/api/history/${scenario_id}`;

    const res = await (await fetch(endpoint)).json();

    if (res.error) {
      throw res;
    }

    const { history } = res;

    dispatch({ type: GET_RUN_HISTORY_SUCCESS, history });
    return { ...history };
  } catch (error) {
    dispatch({ type: GET_RUN_HISTORY_ERROR, error });
    return null;
  }
};
