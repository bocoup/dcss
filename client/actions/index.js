import {
    LOG_IN,
    LOG_OUT,
    SET_RUN,
    SET_SCENARIO,
    SET_SCENARIOS,
    SET_SLIDES,
    SET_USERS
} from './types';

export const logIn = username => ({
    type: LOG_IN,
    payload: {
        username,
        isLoggedIn: true
    }
});

export const logOut = username => ({
    type: LOG_OUT,
    payload: {
        username,
        isLoggedIn: false
    }
});

export const setRun = payload => ({
    type: SET_RUN,
    payload
});

export const setScenario = payload => ({
    type: SET_SCENARIO,
    payload
});

export const setScenarios = payload => ({
    type: SET_SCENARIOS,
    payload
});

export const setSlides = payload => ({
    type: SET_SLIDES,
    payload
});

export const setUsers = payload => ({
    type: SET_USERS,
    payload
});
