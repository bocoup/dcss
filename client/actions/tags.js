import { GET_CATEGORIES_SUCCESS, GET_CATEGORIES_ERROR } from './types';

export const getCategories = () => async dispatch => {
  try {
    const res = await (await fetch('/api/tags/categories')).json();

    if (res.error) {
      throw res;
    }

    const {
      categories
    } = res;

    dispatch({ type: GET_CATEGORIES_SUCCESS, categories });
    return categories;
  } catch (error) {
    dispatch({ type: GET_CATEGORIES_ERROR, error });
    return null;
  }
};
