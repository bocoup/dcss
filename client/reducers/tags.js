import { combineReducers } from 'redux';

import {
  GET_CATEGORIES_SUCCESS,
} from '@client/actions/types';

export const categories = (state = [], action) => {
  const { categories = [], type } = action;
  switch (type) {
    case GET_CATEGORIES_SUCCESS:
      return categories;
    default:
      return state;
  }
};

export default combineReducers({
  categories
});
