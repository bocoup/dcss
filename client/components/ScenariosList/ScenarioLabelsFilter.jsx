import React from 'react';
import PropTypes from 'prop-types';
import * as QueryString from 'query-string';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { Dropdown } from '@components/UI';
import { setLabelsInUse } from '@actions/tags';
import Identity from '@utils/Identity';

const qsOpts = {
  arrayFormat: 'bracket'
};

function makeQueryString(labels) {
  const qs = {};
  const { search } = QueryString.parse(window.location.search, qsOpts);

  if (search) {
    qs.search = search;
  }

  if (labels && labels.length) {
    qs.labels = labels;
  }

  return `?${QueryString.stringify(qs, qsOpts)}`;
}

class ScenarioLabelsFilter extends React.Component {
  constructor(props) {
    super(props);
    this.onClick = this.onClick.bind(this);
  }

  onClick(event, { value }) {
    event.stopPropagation();

    const labelsInUse = this.props.tags.labelsInUse.slice(0);

    if (labelsInUse.includes(value)) {
      labelsInUse.splice(labelsInUse.indexOf(value), 1);
    } else {
      labelsInUse.push(value);
    }

    this.props.setLabelsInUse(labelsInUse);
    this.props.history.push(
      `${this.props.location.pathname}${makeQueryString(labelsInUse)}`
    );
  }

  render() {
    const { onClick } = this;
    return (
      <Dropdown item text="Filter by labels" closeOnChange={false}>
        <Dropdown.Menu>
          {this.props.tags.labels.map(label => {
            const key = Identity.key(label);
            const active = this.props.tags.labelsInUse.includes(label.value);
            const itemProps = {
              ...label,
              active,
              'aria-label': label.text
            };
            return (
              <Dropdown.Item {...itemProps} key={key} onClick={onClick}>
                {label.text}
              </Dropdown.Item>
            );
          })}
        </Dropdown.Menu>
      </Dropdown>
    );
  }
}

ScenarioLabelsFilter.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  tags: PropTypes.shape({
    labels: PropTypes.array,
    labelsInUse: PropTypes.array
  }),
  location: PropTypes.object,
  setLabelsInUse: PropTypes.func
};

const mapStateToProps = state => {
  const { tags } = state;
  return { tags };
};

const mapDispatchToProps = dispatch => ({
  setLabelsInUse: params => dispatch(setLabelsInUse(params))
});

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(ScenarioLabelsFilter)
);
