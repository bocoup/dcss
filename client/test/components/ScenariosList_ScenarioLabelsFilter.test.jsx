import React from 'react';
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useLayoutEffect: jest.requireActual('react').useEffect
}));

import {
  fetchImplementation,
  mounter,
  reduxer,
  serialize,
  snapshotter,
  state
} from '../bootstrap';
import { unmountComponentAtNode } from 'react-dom';

import {
  act,
  fireEvent,
  prettyDOM,
  render,
  screen,
  waitFor
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { SET_LABELS_IN_USE_SUCCESS } from '../../actions/types';
import * as tagsActions from '../../actions/tags';
jest.mock('../../actions/tags');

import ScenarioLabelsFilter from '../../components/ScenariosList/ScenarioLabelsFilter.jsx';

const original = JSON.parse(JSON.stringify(state));
let container = null;
let commonProps = null;
let commonState = null;

beforeAll(() => {
  (window || global).fetch = jest.fn();
});

afterAll(() => {
  jest.restoreAllMocks();
});

beforeEach(() => {
  jest.useFakeTimers();
  container = document.createElement('div');
  container.setAttribute('id', 'root');
  document.body.appendChild(container);

  fetchImplementation(fetch);

  tagsActions.setLabelsInUse = jest.fn();
  tagsActions.setLabelsInUse.mockImplementation(
    labelsInUse => async dispatch => {
      dispatch({ type: SET_LABELS_IN_USE_SUCCESS, labelsInUse });
      return labelsInUse;
    }
  );

  commonProps = {};
  commonState = JSON.parse(JSON.stringify(original));
});

afterEach(() => {
  jest.runOnlyPendingTimers();
  jest.useRealTimers();
  jest.resetAllMocks();
  unmountComponentAtNode(container);
  container.remove();
  container = null;
  commonProps = null;
  commonState = null;
});

test('ScenarioLabelsFilter', () => {
  expect(ScenarioLabelsFilter).toBeDefined();
});

test('Render 1 1', async done => {
  const Component = ScenarioLabelsFilter;

  const props = {
    ...commonProps
  };

  const state = {
    ...commonState
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);

  const { asFragment } = render(<ConnectedRoutedComponent {...props} />);
  expect(asFragment()).toMatchSnapshot();

  userEvent.click(await screen.findByRole('listbox'));
  expect(asFragment()).toMatchSnapshot();

  userEvent.click(await screen.findByRole('listbox'));
  expect(asFragment()).toMatchSnapshot();

  done();
});
