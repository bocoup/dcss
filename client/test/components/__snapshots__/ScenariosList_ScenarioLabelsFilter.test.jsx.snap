// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render 1 1 1`] = `
<DocumentFragment>
  <div
    aria-expanded="false"
    class="ui item dropdown"
    role="listbox"
    tabindex="0"
  >
    <div
      aria-atomic="true"
      aria-live="polite"
      class="divider text"
      role="alert"
    >
      Filter by labels
    </div>
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition"
    >
      <div
        aria-checked="false"
        class="item"
        id="3"
        name="fun"
        role="option"
      />
      <div
        aria-checked="false"
        class="item"
        id="5"
        name="Easy"
        role="option"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Render 1 1 2`] = `
<DocumentFragment>
  <div
    aria-expanded="true"
    class="ui active visible item dropdown"
    role="listbox"
    tabindex="0"
  >
    <div
      aria-atomic="true"
      aria-live="polite"
      class="divider text"
      role="alert"
    >
      Filter by labels
    </div>
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition visible"
    >
      <div
        aria-checked="false"
        class="item"
        id="3"
        name="fun"
        role="option"
      />
      <div
        aria-checked="false"
        class="item"
        id="5"
        name="Easy"
        role="option"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Render 1 1 3`] = `
<DocumentFragment>
  <div
    aria-expanded="false"
    class="ui item dropdown"
    role="listbox"
    tabindex="0"
  >
    <div
      aria-atomic="true"
      aria-live="polite"
      class="divider text"
      role="alert"
    >
      Filter by labels
    </div>
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition"
    >
      <div
        aria-checked="false"
        class="item"
        id="3"
        name="fun"
        role="option"
      />
      <div
        aria-checked="false"
        class="item"
        id="5"
        name="Easy"
        role="option"
      />
    </div>
  </div>
</DocumentFragment>
`;
